### ISSUE TITLE
Questions from a beginner

### ISSUE THREAD

**Issue by gauteh** [Stoeoef/rstar#16]
Hi, I'm a little confused about r*-tree terminology, but as far as I can see from demo and so on by adding points the rtree-boxes/rectangles are automatically created? That is, I do not need to implement the splitting algorithm myself?

**Comment by urschrei** [comment]
You mean when constructing the tree? That's correct: you construct it using your input points, and then you can query the resulting `RTree`.

**Comment by gauteh** [comment]
Great, thanks :)

**Comment by gauteh** [comment]
Another question; I am trying to determine whether points are covered by (any) polygon. I've built my tree, and I've implemented PointsDistance for my polygon type (or a wrapper of it). Does `locate_at_point` return a polygon only if the point is inside the polygon or, also if the point is inside the rectangle. Meaning I also need to check if the point is inside the returned polygon.

**Comment by Stoeoef** [comment]
`locate_at_point` will use the polygon's AABB (rectangle) only to determine a _list of possible intersection candidates_. For each rectangle in this candidate list, [contains_point](https://docs.rs/rstar/0.5.1/rstar/trait.PointDistance.html#method.contains_point) is then called to check if the point is really inside the polygon. The default implementation will check if the `distance_2` is less or equal than zero. As calculating the distance² is usually more expensive than checking for containment, it can be advised to overwrite `contains_point` for performance reasons.

TL; DR: `contains_point` does return only objects which "really" contain the point if trait `PointDistance` is implemented properly.

If you need to get all objects in a rectangle, use [locate_in_envelope](https://docs.rs/rstar/0.5.1/rstar/struct.RTree.html#method.locate_in_envelope).

I hope that answers your question!

**Comment by gauteh** [comment]
Yes, I think that clarifies it :) I implemented `contains_point` and that sped things up. That means I do not need to additionally check against the polygon.

My use case is checking a large number of points (300.000, but lets start with 60k for the `high`-resolutions dataset with about 140k polygons once they have been joined together at overlapping exteriors) against land shapes from [GSHHS](https://www.ngdc.noaa.gov/mgg/shorelines/) to determine whether they are on land or in the ocean. This needs to happen pretty quickly as it is used in simulation. I was hoping that with R* I would be able to just work with the entire world and avoid any other dynamic bounding boxing of the simulation space. The data is given in lon/lat so its not exactly ideal for the geometry used here, have not tried to re-project it for the R*-tree. Anyway, GEOS contain checks through Shapely (python) seem to be vectorized and still perform about ten times better than my implementation using rstar. I am now thinking about a way to vectorize this for an array/slice of input points to check using `locate_at_point` against the R*-tree, not yet sure if it is possible -- or if it will only be possible by effectively checking against all envelopes and then checking against all polygons in the matching envelopes, this might still be faster.

**Comment by urschrei** [comment]
> Anyway, GEOS contain checks through Shapely (python) seem to be vectorized and still perform about ten times better than my implementation using rstar

That seems like a big difference. Could you post some code so we can see how you're doing it in Python vs Rust? A sample of the points and a sample of stitched-together polygons as GeoJSON if you can – enough to show a timing difference, essentially.

**Comment by gauteh** [comment]
Yep, I'll tidy it up a bit tomorrow.

tir. 8. okt. 2019, 22:49 skrev Stephan Hügel <notifications@github.com>:

**Comment by gauteh** [comment]
Here are my benchmarks: https://github.com/gauteh/landmask/tree/master/landmaskpy. It is built using [maturin](https://github.com/PyO3/maturin). The rust benchmarks are kind of tedious to run since the Bencher runs a minimum of iterations. Easiest way to test:

```sh
$ cd landmask/landmaskpy
$ pip install -r requirements-dev.txt # you have preferably already set up your pyenv here
$ maturin develop --release # you might have to specify interpreter using e.g. `-i python3.6`
$ pytest
```

**Comment by Stoeoef** [comment]
I've had a look at your implementation and the profile. I've noticed two points:
1. You can get significant (about factor 7 for me) speedups for the r-tree creation if you cache a polygon's AABB. Add it as another field to your polygon wrapper.

_Note_: The above factor only represents the R-Tree creation speedup. This makes up only a small part of loading a .wkb file. The total speed up of loading a .wkb file will be smaller. Also, this solution requires more memory.

2. If you change `PolyWrapper::contains_point` to 
```Rust
    fn contains_point(&self, point: &[f64; 2]) -> bool {
        let p = geo::Point::new(point[0], point[1]);
        self.1.contains_point(point) // new, self.1 is the cached AABB
           && self.0.contains(&p)
    }
```
 "contains" can be improved by nearly factor 2 as it seems. Admittedly, I thought that `contains_point` would only be called if the AABB can potentially contain the point, but I was wrong. I might add this in future versions of rstar.
With the above change implemented, about 68% of the time seems to be spent within [`get_position`](https://github.com/georust/geo/blob/d647d86dd56b3acb4acddb6828f0919ad19f79ec/geo/src/algorithm/contains.rs#L123). `get_position` calls [`line_string_contains_point`](https://github.com/georust/geo/blob/cae9460b1440b84d2ff6f92418f581c12536f54d/geo-types/src/private_utils.rs#L130), which takes about 28% of total time. The latter is only required if a point lies exactly on the boundary, if that is not important, you could try to optimize `contains_point` accordingly.

I haven't yet have a look at Shapely, maybe I'll find some time to do. I'll let you know what I find!

**Comment by gauteh** [comment]
Thanks for the extensive answer! In my case I can also skip checking interior lines. The GEOS library and Shapely has some nice optimizations for stuff like `contains` etc, they also have `prepared geometries` which are similar to storing `AABB` in rstar's case. I think that by maybe making a special `contains` method taking slices of points we might be able to propagate the loop far enough down the line so that rust might be able to vectorize it. Looking at r-star's code it probably requires a new/modified locator struct. Since rstars `locate_at_point` fully determines whether it is contained as well it will probably be pretty close to this behaviour.

For the project that I started looking into this I have had very good luck with a 500mx500m rasterized grid which I use to mask points which are not contained in any points + shapely vectorized on the remaining points against the entire set of polygons (or `shapely.vectorized.contains`), theoretically putting in an r-tree before the polygon check should drastically reduce the number of polygons.. but still the problem of vectorization remains since the points could be contained in any polygon of the total set.

**Comment by gauteh** [comment]
Tried out your optimizations. Works pretty well for the coarse set.
